importHelpers with symlinks breaks type resolution

<!-- 🚨 STOP 🚨 𝗦𝗧𝗢𝗣 🚨 𝑺𝑻𝑶𝑷 🚨

Half of all issues filed here are duplicates, answered in the FAQ, or not appropriate for the bug tracker. Even if you think you've found a *bug*, please read the FAQ first, especially the Common "Bugs" That Aren't Bugs section!

Please help us by doing the following steps before logging an issue:
  * Search: https://github.com/Microsoft/TypeScript/search?type=Issues
  * Read the FAQ: https://github.com/Microsoft/TypeScript/wiki/FAQ

Please fill in the *entire* template below.
-->

<!-- Please try to reproduce the issue with `typescript@next`. It may have already been fixed. -->
**TypeScript Version:**  3.2.2

<!-- Search terms you tried before logging this (so others can find this issue more easily) -->
**Search Terms:**
type symlink importHelpers pnpm

**Code**
https://github.com/stevefan1999-personal/gqlify/tree/patch-pnpm-tslib

**Reproduction:**
This is a little bit complicated, you will need to clone the repo I provided and checkout the branch, then run `pnpm install`, and then please flip the `importHelpers` options to false and re-run the installation again.

**Expected behavior:**
When `importHelpers` is true, it shouldn't break type inference like we should when `importHelpers` is false.

**Actual behavior:**
```
error TS2742: The inferred type of '...' cannot be named without a reference to '...'. This is likely not portable. A type annotation is necessary.
```
When `importHelpers` is true and the @types directory is a symlink.

**Related Issues:**
https://github.com/pnpm/pnpm/issues/1375

**Extra Stuff:**
Although it is application-specific (Lerna with NPM/Yarn works fine on this), I do think this should be handled correctly regardless of any layers of symlinks.

PS: 
tsc output observation diff: http://www.mergely.com/a6N4kOW3/

Similar problem here and I guess the reason is the same:

```
ERROR in components/empty/nz-embed-empty.component.ts(36,3): error TS2742: The inferred type of 'defaultSvg' cannot be named without a reference to 'D:/Developer/ng-zorro-antd/node_modules/@angular/platform-browser/platform-browser'. This is likely not portable. A type annotation is necessary.
components/empty/nz-empty.component.ts(34,3): error TS2742: The inferred type of 'defaultSvg' cannot be named without a reference to 'D:/Developer/ng-zorro-antd/node_modules/@angular/platform-browser/platform-browser'. This is likely not portable. A type annotation is necessary.
components/select/nz-select.service.ts(45,3): error TS2742: The inferred type of 'open$' cannot be named without a reference to 'D:/Developer/ng-zorro-antd/node_modules/rxjs'. This is likely not portable. A type annotation is necessary.
components/select/nz-select.service.ts(52,3): error TS2742: The inferred type of 'listOfSelectedValue$' cannot be named without a reference to 'D:/Developer/ng-zorro-antd/node_modules/rxjs'. This is likely not portable. A type annotation is necessary.
components/select/nz-select.service.ts(53,3): error TS2742: The inferred type of 'modelChange$' cannot be named without a reference to 'D:/Developer/ng-zorro-antd/node_modules/rxjs'. This is likely not portable. A type annotation is necessary.
components/select/nz-select.service.ts(69,3): error TS2742: The inferred type of 'searchValue$' cannot be named without a reference to 'D:/Developer/ng-zorro-antd/node_modules/rxjs'. This is likely not portable. A type annotation is necessary.
components/select/nz-select.service.ts(97,3): error TS2742: The inferred type of 'valueOrOption$' cannot be named without a reference to 'D:/Developer/ng-zorro-antd/node_modules/rxjs'. This is likely not portable. A type annotation is necessary.
components/select/nz-select.service.ts(113,3): error TS2742: The inferred type of 'check$' cannot be named without a reference to 'D:/Developer/ng-zorro-antd/node_modules/rxjs'. This is likely not portable. A type annotation is necessary.
```

I ran into this issue as well. My project was built with Bazel. 

A few snippets to reproduce the error:
```ts
import {Parser as ParserBase} from 'chevrotain'


const parser = createParser()

// @ts-ignore: TS2742
const BaseCstVisitor = parser.getBaseCstVisitorConstructor();

function createParser(): Parser {
    return new Parser()
}

class Parser extends ParserBase {
...
}
```

The compilation error after upgrading from ts `3.1.3` to `3.2.4`:
```text
src/ui/lib/dsl/langs/logi/semantic.ts:13:7 - error TS2742: The inferred type of 'BaseCstVisitor' cannot be named without a reference to '../../../../../../external/npm/node_modules/chevrotain/lib/chevrotain'. This is likely not portable. A type annotation is necessary.

13 const BaseCstVisitor = parser.getBaseCstVisitorConstructor();
         ~~~~~~~~~~~~~~
```

What I want is the ability to just mute this TS2742 error and make the code compile again.

I have tried adding `// @ts-ignore`, no luck.

Help needed. Thanks in advance!

--- a/src/compiler/moduleSpecifiers.ts
+++ b/src/compiler/moduleSpecifiers.ts
@@ -175,9 +175,9 @@ namespace ts.moduleSpecifiers {
 
     function discoverProbableSymlinks(files: ReadonlyArray<SourceFile>, getCanonicalFileName: GetCanonicalFileName, cwd: string): ReadonlyMap<string> {
         const result = createMap<string>();
-        const symlinks = mapDefined(files, sf =>
-            sf.resolvedModules && firstDefinedIterator(sf.resolvedModules.values(), res =>
-                res && res.originalPath && res.resolvedFileName !== res.originalPath ? [res.resolvedFileName, res.originalPath] : undefined));
+        const symlinks = flatten<readonly [string, string]>(mapDefined(files, sf =>
+            sf.resolvedModules && compact(arrayFrom(mapIterator(sf.resolvedModules.values(), res =>
+                res && res.originalPath && res.resolvedFileName !== res.originalPath ? [res.resolvedFileName, res.originalPath] as const : undefined)))));
         for (const [resolvedPath, originalPath] of symlinks) {
             const [commonResolved, commonOriginal] = guessDirectorySymlink(resolvedPath, originalPath, cwd, getCanonicalFileName);
             result.set(commonOriginal, commonResolved);

--- a/tests/baselines/reference/declarationEmitForGlobalishSpecifierSymlink.js
+++ b/tests/baselines/reference/declarationEmitForGlobalishSpecifierSymlink.js
@@ -1,0 +1,45 @@
+//// [tests/cases/compiler/declarationEmitForGlobalishSpecifierSymlink.ts] ////
+
+//// [impl.d.ts]
+export function getA(): A;
+export enum A {
+    Val
+}
+//// [index.d.ts]
+export * from "./src/impl";
+//// [package.json]
+{
+    "name": "typescript-fsa",
+    "version": "1.0.0"
+}
+//// [impl.d.ts]
+export function getA(): A;
+export enum A {
+    Val
+}
+//// [index.d.ts]
+export * from "./src/impl";
+//// [package.json]
+{
+    "name": "typescript-fsa",
+    "version": "1.0.0"
+}
+//// [index.ts]
+import * as _whatever from "p2";
+import { getA } from "typescript-fsa";
+
+export const a = getA();
+//// [index.d.ts]
+export const a: import("typescript-fsa").A;
+
+
+
+//// [index.js]
+"use strict";
+exports.__esModule = true;
+var typescript_fsa_1 = require("typescript-fsa");
+exports.a = typescript_fsa_1.getA();
+
+
+//// [index.d.ts]
+export declare const a: import("typescript-fsa").A;

--- a/tests/baselines/reference/declarationEmitForGlobalishSpecifierSymlink.symbols
+++ b/tests/baselines/reference/declarationEmitForGlobalishSpecifierSymlink.symbols
@@ -1,0 +1,43 @@
+=== /p1/node_modules/typescript-fsa/src/impl.d.ts ===
+export function getA(): A;
+>getA : Symbol(getA, Decl(impl.d.ts, 0, 0))
+>A : Symbol(A, Decl(impl.d.ts, 0, 26))
+
+export enum A {
+>A : Symbol(A, Decl(impl.d.ts, 0, 26))
+
+    Val
+>Val : Symbol(A.Val, Decl(impl.d.ts, 1, 15))
+}
+=== /p1/node_modules/typescript-fsa/index.d.ts ===
+export * from "./src/impl";
+No type information for this code.=== /p2/node_modules/typescript-fsa/src/impl.d.ts ===
+export function getA(): A;
+>getA : Symbol(getA, Decl(impl.d.ts, 0, 0))
+>A : Symbol(A, Decl(impl.d.ts, 0, 26))
+
+export enum A {
+>A : Symbol(A, Decl(impl.d.ts, 0, 26))
+
+    Val
+>Val : Symbol(A.Val, Decl(impl.d.ts, 1, 15))
+}
+=== /p2/node_modules/typescript-fsa/index.d.ts ===
+export * from "./src/impl";
+No type information for this code.=== /p1/index.ts ===
+import * as _whatever from "p2";
+>_whatever : Symbol(_whatever, Decl(index.ts, 0, 6))
+
+import { getA } from "typescript-fsa";
+>getA : Symbol(getA, Decl(index.ts, 1, 8))
+
+export const a = getA();
+>a : Symbol(a, Decl(index.ts, 3, 12))
+>getA : Symbol(getA, Decl(index.ts, 1, 8))
+
+=== /p2/index.d.ts ===
+export const a: import("typescript-fsa").A;
+>a : Symbol(a, Decl(index.d.ts, 0, 12))
+>A : Symbol(A, Decl(impl.d.ts, 0, 26))
+
+

--- a/tests/baselines/reference/declarationEmitForGlobalishSpecifierSymlink.types
+++ b/tests/baselines/reference/declarationEmitForGlobalishSpecifierSymlink.types
@@ -1,0 +1,41 @@
+=== /p1/node_modules/typescript-fsa/src/impl.d.ts ===
+export function getA(): A;
+>getA : () => A
+
+export enum A {
+>A : A
+
+    Val
+>Val : A
+}
+=== /p1/node_modules/typescript-fsa/index.d.ts ===
+export * from "./src/impl";
+No type information for this code.=== /p2/node_modules/typescript-fsa/src/impl.d.ts ===
+export function getA(): A;
+>getA : () => A
+
+export enum A {
+>A : A
+
+    Val
+>Val : A
+}
+=== /p2/node_modules/typescript-fsa/index.d.ts ===
+export * from "./src/impl";
+No type information for this code.=== /p1/index.ts ===
+import * as _whatever from "p2";
+>_whatever : typeof _whatever
+
+import { getA } from "typescript-fsa";
+>getA : () => import("/p1/node_modules/typescript-fsa/index").A
+
+export const a = getA();
+>a : import("/p1/node_modules/typescript-fsa/index").A
+>getA() : import("/p1/node_modules/typescript-fsa/index").A
+>getA : () => import("/p1/node_modules/typescript-fsa/index").A
+
+=== /p2/index.d.ts ===
+export const a: import("typescript-fsa").A;
+>a : import("/p2/node_modules/typescript-fsa/index").A
+
+

--- a/tests/baselines/reference/declarationEmitForGlobalishSpecifierSymlink2.js
+++ b/tests/baselines/reference/declarationEmitForGlobalishSpecifierSymlink2.js
@@ -1,0 +1,33 @@
+//// [tests/cases/compiler/declarationEmitForGlobalishSpecifierSymlink2.ts] ////
+
+//// [impl.d.ts]
+export function getA(): A;
+export enum A {
+    Val
+}
+//// [index.d.ts]
+export * from "./src/impl";
+//// [package.json]
+{
+    "name": "typescript-fsa",
+    "version": "1.0.0"
+}
+//// [index.ts]
+import * as _whatever from "p2";
+import { getA } from "typescript-fsa";
+
+export const a = getA();
+//// [index.d.ts]
+export const a: import("typescript-fsa").A;
+
+
+
+//// [index.js]
+"use strict";
+exports.__esModule = true;
+var typescript_fsa_1 = require("typescript-fsa");
+exports.a = typescript_fsa_1.getA();
+
+
+//// [index.d.ts]
+export declare const a: import("typescript-fsa").A;

--- a/tests/baselines/reference/declarationEmitForGlobalishSpecifierSymlink2.symbols
+++ b/tests/baselines/reference/declarationEmitForGlobalishSpecifierSymlink2.symbols
@@ -1,0 +1,30 @@
+=== /cache/typescript-fsa/src/impl.d.ts ===
+export function getA(): A;
+>getA : Symbol(getA, Decl(impl.d.ts, 0, 0))
+>A : Symbol(A, Decl(impl.d.ts, 0, 26))
+
+export enum A {
+>A : Symbol(A, Decl(impl.d.ts, 0, 26))
+
+    Val
+>Val : Symbol(A.Val, Decl(impl.d.ts, 1, 15))
+}
+=== /cache/typescript-fsa/index.d.ts ===
+export * from "./src/impl";
+No type information for this code.=== /p1/index.ts ===
+import * as _whatever from "p2";
+>_whatever : Symbol(_whatever, Decl(index.ts, 0, 6))
+
+import { getA } from "typescript-fsa";
+>getA : Symbol(getA, Decl(index.ts, 1, 8))
+
+export const a = getA();
+>a : Symbol(a, Decl(index.ts, 3, 12))
+>getA : Symbol(getA, Decl(index.ts, 1, 8))
+
+=== /p2/index.d.ts ===
+export const a: import("typescript-fsa").A;
+>a : Symbol(a, Decl(index.d.ts, 0, 12))
+>A : Symbol(A, Decl(impl.d.ts, 0, 26))
+
+

--- a/tests/baselines/reference/declarationEmitForGlobalishSpecifierSymlink2.types
+++ b/tests/baselines/reference/declarationEmitForGlobalishSpecifierSymlink2.types
@@ -1,0 +1,29 @@
+=== /cache/typescript-fsa/src/impl.d.ts ===
+export function getA(): A;
+>getA : () => A
+
+export enum A {
+>A : A
+
+    Val
+>Val : A
+}
+=== /cache/typescript-fsa/index.d.ts ===
+export * from "./src/impl";
+No type information for this code.=== /p1/index.ts ===
+import * as _whatever from "p2";
+>_whatever : typeof _whatever
+
+import { getA } from "typescript-fsa";
+>getA : () => import("/cache/typescript-fsa/index").A
+
+export const a = getA();
+>a : import("/cache/typescript-fsa/index").A
+>getA() : import("/cache/typescript-fsa/index").A
+>getA : () => import("/cache/typescript-fsa/index").A
+
+=== /p2/index.d.ts ===
+export const a: import("typescript-fsa").A;
+>a : import("/cache/typescript-fsa/index").A
+
+

--- a/tests/cases/compiler/declarationEmitForGlobalishSpecifierSymlink.ts
+++ b/tests/cases/compiler/declarationEmitForGlobalishSpecifierSymlink.ts
@@ -1,0 +1,35 @@
+// @useCaseSensitiveFilenames: true
+// @declaration: true
+// @filename: /p1/node_modules/typescript-fsa/src/impl.d.ts
+export function getA(): A;
+export enum A {
+    Val
+}
+// @filename: /p1/node_modules/typescript-fsa/index.d.ts
+export * from "./src/impl";
+// @filename: /p1/node_modules/typescript-fsa/package.json
+{
+    "name": "typescript-fsa",
+    "version": "1.0.0"
+}
+// @filename: /p2/node_modules/typescript-fsa/src/impl.d.ts
+export function getA(): A;
+export enum A {
+    Val
+}
+// @filename: /p2/node_modules/typescript-fsa/index.d.ts
+export * from "./src/impl";
+// @filename: /p2/node_modules/typescript-fsa/package.json
+{
+    "name": "typescript-fsa",
+    "version": "1.0.0"
+}
+// @filename: /p1/index.ts
+import * as _whatever from "p2";
+import { getA } from "typescript-fsa";
+
+export const a = getA();
+// @filename: /p2/index.d.ts
+export const a: import("typescript-fsa").A;
+
+// @link: /p2 -> /p1/node_modules/p2

--- a/tests/cases/compiler/declarationEmitForGlobalishSpecifierSymlink2.ts
+++ b/tests/cases/compiler/declarationEmitForGlobalishSpecifierSymlink2.ts
@@ -1,0 +1,25 @@
+// @useCaseSensitiveFilenames: true
+// @declaration: true
+// @filename: /cache/typescript-fsa/src/impl.d.ts
+export function getA(): A;
+export enum A {
+    Val
+}
+// @filename: /cache/typescript-fsa/index.d.ts
+export * from "./src/impl";
+// @filename: /cache/typescript-fsa/package.json
+{
+    "name": "typescript-fsa",
+    "version": "1.0.0"
+}
+// @filename: /p1/index.ts
+import * as _whatever from "p2";
+import { getA } from "typescript-fsa";
+
+export const a = getA();
+// @filename: /p2/index.d.ts
+export const a: import("typescript-fsa").A;
+
+// @link: /p2 -> /p1/node_modules/p2
+// @link: /cache/typescript-fsa -> /p1/node_modules/typescript-fsa
+// @link: /cache/typescript-fsa -> /p2/node_modules/typescript-fsa

